File: workout/business/src/domain/workout.rs
use crate::commons::entity_mapper::EntityMapper;
use crate::commons::functions::{string_to_uuid, uuid_to_string};
pub use crate::domain::enums::Visibility;
use crate::domain::exercise::Exercise;
use chrono::NaiveDateTime;
use entity::workout_entity::{ActiveModel, WorkoutEntity};
use sea_orm::{NotSet, Set};
use crate::domain::enums::Difficulty;

#[derive(Debug, Clone)]
pub struct Workout {
    pub id: Option<i32>,
    pub uuid: Option<String>,
    pub owner_id: i32,
    pub owner_uuid: String,
    pub name: String,
    pub description: Option<String>,
    pub difficulty: Difficulty,
    pub muscle_group: String,
    pub exercises: Vec<Exercise>,
    pub visibility: Visibility,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
}

pub struct WorkoutEntityMapper {}

impl EntityMapper<Workout, WorkoutEntity, ActiveModel> for WorkoutEntityMapper {
    fn build_active_model(d: Workout) -> ActiveModel {
        ActiveModel {
            id: match d.id {
                Some(id) => Set(id),
                None => NotSet,
            },
            uuid: match d.uuid {
                Some(uuid) => Set(string_to_uuid(&uuid)),
                None => NotSet,
            },
            owner_id: Set(d.owner_id),
            owner_uuid: Set(string_to_uuid(&d.owner_uuid)),
            name: Set(d.name),
            description: Set(d.description),
            difficulty: Set(d.difficulty.to_string()),
            muscle_group: Set(d.muscle_group),
            visibility: Set(d.visibility.to_string()),
            created_at: NotSet,
            updated_at: NotSet,
        }
    }

    fn from_model(e: WorkoutEntity) -> Workout {
        Workout {
            id: Some(e.id),
            name: e.name,
            description: e.description,
            difficulty: Difficulty::from_string(e.difficulty.as_str()),
            muscle_group: e.muscle_group,
            owner_id: e.owner_id,
            owner_uuid: uuid_to_string(e.owner_uuid),
            exercises: Vec::new(),
            uuid: Some(uuid_to_string(e.uuid)),
            created_at: Some(e.created_at.naive_utc()),
            updated_at: Some(e.updated_at.naive_utc()),
            visibility: Visibility::from_string(e.visibility.as_str()),
        }
    }

    fn from_active_model(e: ActiveModel) -> Workout {
        Workout {
            id: Some(e.id.unwrap()),
            name: e.name.unwrap(),
            description: e.description.unwrap(),
            difficulty: Difficulty::from_string(e.difficulty.unwrap().as_str()),
            muscle_group: e.muscle_group.unwrap(),
            owner_id: e.owner_id.unwrap(),
            owner_uuid: uuid_to_string(e.owner_uuid.unwrap()),
            exercises: Vec::new(),
            uuid: Some(uuid_to_string(e.uuid.unwrap())),
            created_at: Some(e.created_at.unwrap().naive_utc()),
            updated_at: Some(e.updated_at.unwrap().naive_utc()),
            visibility: Visibility::from_string(e.visibility.unwrap().as_str()),
        }
    }
}

File: socialgym_mobile/lib/models/workout.dart
import 'exercise.dart';

class Workout {
  final int? id;
  final String? uuid;
  final int ownerId;
  final String ownerUuid;
  final String name;
  final String? description;
  final String difficulty;
  final String muscleGroup;
  final String visibility;
  final List<Exercise> exercises;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Workout({
    this.id,
    this.uuid,
    required this.ownerId,
    required this.ownerUuid,
    required this.name,
    this.description,
    this.difficulty = 'soft',
    this.muscleGroup = '',
    this.visibility = 'Private',
    this.exercises = const [],
    this.createdAt,
    this.updatedAt,
  });

  factory Workout.fromJson(Map<String, dynamic> json) {
    return Workout(
      id: json['id'],
      uuid: json['uuid'],
      ownerId: json['ownerId'],
      ownerUuid: json['ownerUuid'],
      name: json['name'] ?? '',
      description: json['description'],
      difficulty: json['difficulty'] ?? 'soft',
      muscleGroup: json['muscleGroup'] ?? '',
      visibility: json['visibility'] ?? 'Private',
      exercises:
          (json['exercises'] as List<dynamic>?)
              ?.map((e) => Exercise.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'uuid': uuid,
      'ownerId': ownerId,
      'ownerUuid': ownerUuid,
      'name': name,
      'description': description,
      'difficulty': difficulty,
      'muscleGroup': muscleGroup,
      'visibility': visibility,
      'exercises': exercises.map((e) => e.toJson()).toList(),
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  List<String> get muscleGroups => muscleGroup.isEmpty ? [] : muscleGroup.split('|');

  Workout copyWith({
    int? id,
    String? uuid,
    String? name,
    String? description,
    String? difficulty,
    String? muscleGroup,
    String? visibility,
    int? ownerId,
    String? ownerUuid,
    List<Exercise>? exercises,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Workout(
      id: id ?? this.id,
      uuid: uuid ?? this.uuid,
      name: name ?? this.name,
      description: description ?? this.description,
      difficulty: difficulty ?? this.difficulty,
      muscleGroup: muscleGroup ?? this.muscleGroup,
      visibility: visibility ?? this.visibility,
      ownerId: ownerId ?? this.ownerId,
      ownerUuid: ownerUuid ?? this.ownerUuid,
      exercises: exercises ?? this.exercises,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
